### PR TITLE
fix(tree): maintain anchors through transaction abort

### PR DESCRIPTION
**Pull request by yann-achard-MS**
This PR fixes a bug where anchors were not properly updated in the face of transaction abort.

### CODE REVIEW

File: packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts [hunk 1]
```diff
@@ -1017,6 +1036,62 @@ describe("SharedTree", () => {
 			};
 			assert(compareUpPaths(childPath, expected));
 		});
+
+		it("Anchors are adjusted in the face of edits", async () => {
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts:1040]: Why is a test for editManager's anchor handling depending on Fluid, sharedTree, localServer etc.?

EditManager is in core. We should be able to test it without any of this stuff. This bug can be reproduced with a trivial edit family that does nothing but add and subtract from a global number. It should need modular change family, much less a full collaborating fluid server implementation.

If you are going to use a full shared tree instance (and you shouldn't) at least use the mock runtime. There is an example of doing that over in https://github.com/microsoft/FluidFramework/pull/14608/files#diff-9cdbb4a83063ab89dbffdc8570428081139e7ab2ac76201e9eb6202c808bc006R25

Comment (CraigMacomber) [reply packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts:1040]: Having an integration test may be useful, but going all the way up to sharedtree.spec.ts should really be a last resort, as should pulling in a full server implementation.

Comment (yann-achard-MS) [reply packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts:1040]: You're quite right. I've fallen into the habit of adding some tests here and didn't stop to think this was unit-testable.

I wasn't aware there were two runtimes to choose from when it comes to our tests. I take it the one used by `TestTreeProvider` should be reserved for the occasional smoke test? That's kind of surprising to me, I thought `TestTreeProvider` was meant to be the established to write integration tests.
We should probably make it more configurable and use the mock runtime in some of the existing tests in this file.
I also think some of the tests here could be moved over to editing.spec.ts and use whatever the setup is there.

I'll start by writing a unit test for this.

Comment (CraigMacomber) [reply packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts]: Personally, I'd rather see TestTreeProvider broken up into some simpler modular helper functions than add even more dependencies/configuration to it, but as it is, yes, I think it only makes sense for a couple top level smote tests or anything that depends on fluid blob upload or summarization (ideally I'd like to keep such code as unit testable as possible without actually depending on the flid blob stuff, but some integration testing there will likely be important).

In general for unit tests I think we should focus on making it easy to use the various modules on our codebase without having elaborate test environments constructed around them. This includes things like shared-tree core branches: they really should be usable without fluid/sharedTree (have a mock trunk branch).

My default setup is each source files has a test file, and if something is worth testing, write tests for the specific function that contains the test worthy logic, which call that code with as little setup as possible. New tests math this, so I'm approving.

**Review by CraigMacomber — APPROVED** [review]

**Pull request merged**